pinot-query-planner/src/main/java/org/apache/pinot/calcite/rel/rules/PinotWindowExchangeNodeInsertRule.java

```diff
@@ -209,23 +244,37 @@ private void validateWindows(Window window) {
   }
 
   private void validateWindowAggCallsSupported(Window.Group windowGroup) {
-    for (int i = 0; i < windowGroup.aggCalls.size(); i++) {
-      Window.RexWinAggCall aggCall = windowGroup.aggCalls.get(i);
+    for (Window.RexWinAggCall aggCall : windowGroup.aggCalls) {
       SqlKind aggKind = aggCall.getKind();
       Preconditions.checkState(SUPPORTED_WINDOW_FUNCTION_KIND.contains(aggKind),
           String.format("Unsupported Window function kind: %s. Only aggregation functions are supported!", aggKind));
     }
   }
 
   private void validateWindowFrames(Window.Group windowGroup) {
-    // Has ROWS only aggregation call kind (e.g. ROW_NUMBER)?
-    boolean isRowsOnlyTypeAggregateCall = isRowsOnlyAggregationCallType(windowGroup.aggCalls);
+    // Has rank based aggregation call kind (e.g. ROW_NUMBER, RANK, DENSE_RANK)
+    boolean hasRankBasedAgg = hasRankBasedAgg(windowGroup.aggCalls);
+
+    // FOLLOWING lower bound or PRECEDING upper bound is not allowed, and won't reach here
+    RexWindowBound lowerBound = windowGroup.lowerBound;
+    assert !lowerBound.isFollowing();
+    RexWindowBound upperBound = windowGroup.upperBound;
+    assert !upperBound.isPreceding();
+    boolean hasBound = (lowerBound.isPreceding() && !lowerBound.isUnbounded()) || (upperBound.isFollowing()
+        && !upperBound.isUnbounded());
+
+    if (hasRankBasedAgg) {
+      Preconditions.checkState(!hasBound, "Window frame is not supported for ROW_NUMBER, RANK, DENSE_RANK functions");
+    }
+    if (!windowGroup.isRows) {
+      Preconditions.checkState(!hasBound, "Bounded RANGE window frame is currently not supported");
+    }
     // For Phase 1 only the default frame is supported
-    Preconditions.checkState(!windowGroup.isRows || isRowsOnlyTypeAggregateCall,
+    Preconditions.checkState(!windowGroup.isRows || hasRankBasedAgg,
```

Comment:
If it is a rank based aggregation, we already check above that the window frame is unbounded (whether range or rows). So this condition of `!windowGroup.isRows || hasRankBasedAgg` doesn't really seem to make sense here unless I'm missing something?

Reply:
Line 272-283 was supposed to be removed. I accidentally reverted the removal to compare with the new logic